### PR TITLE
multi-section checklists, multiple reviewers and updated git ticket show

### PR DESCRIPTION
**Pull request by dawzee**

### CODE REVIEW

File: bug/snapshot.go [hunk 1]
```diff
@@ -144,17 +144,17 @@ func (snap *Snapshot) HasAnyActor(ids ...entity.Id) bool {
...
 	checklists := make(map[string]Checklist)
 
 	// Only checklists named in the labels list are currently valid
 	for _, l := range snap.Labels {
 		lblStr := string(l)
 		if strings.HasPrefix(lblStr, "checklist:") {
-			var present bool
-			checklists[lblStr], present = snap.Checklists[lblStr]
-			if !present {
+			if snapshotChecklist, present := snap.Checklists[lblStr][reviewer]; present {
```
Comment (aleb) [inline bug/snapshot.go:155]: I see in `GetChecklistCompoundStates` we check that `snap.Checklists[lblStr]` is present.

File: bug/snapshot.go [hunk 2]
```diff
@@ -165,6 +165,37 @@ func (snap *Snapshot) GetChecklists() (map[string]Checklist, error) {
...
+			states[lblStr] = Pending
+
+			clMap, present := snap.Checklists[lblStr]
+			if present {
+				// at least one user has edited this checklist
+				for _, cl := range clMap {
+					clState := cl.CompoundState()
+					if clState == Failed {
+						// someone failed it, it's failed
+						states[lblStr] = Failed
```
Comment (aleb) [inline bug/snapshot.go:187]: When we detect a Failed, we can `break` since it won't get better. Then on the next line we don't have to check for `states[lblStr] != Failed`. The `if` can be converted to a `switch`.

File: commands/show.go [hunk 3]
```diff
@@ -98,12 +126,22 @@ func runShowBug(cmd *cobra.Command, args []string) error {
...
+	// Workflow
+	fmt.Printf("workflow: %s\n", workflow)
+
+	// Checklists
+	fmt.Printf("checklists:\n")
+	for clLabel, st := range snapshot.GetChecklistCompoundStates() {
+		cl, present := bug.ChecklistStore[clLabel]
+
+		if !present {
+			return fmt.Errorf("\nunknown checklist: %s\n", clLabel)
```
Comment (aleb) [inline commands/show.go:138]: There should be no need for \n in error messages.

File: input/input.go [hunk 4]
```diff
@@ -241,51 +246,67 @@ func ChecklistEditorInput(repo repository.RepoCommon, checklist bug.Checklist) (
...
+				commentText = commentText + line + "\n"
 			}
-			commentText = commentText + line + "\n"
 		}
 	}
 
-	if nextQ != len(checklist.Questions)+1 {
-		return checklistChanged, fmt.Errorf("checklist parse error, question count")
+	if nextS != len(checklist.Sections)+1 {
+		// if nextQ != len(checklist.Questions)+1 {
```
Comment (aleb) [inline input/input.go:308]: commented line can be removed

File: input/input.go [hunk 5]
```diff
@@ -241,51 +246,67 @@ func ChecklistEditorInput(repo repository.RepoCommon, checklist bug.Checklist) (
...
 				if err != nil {
 					// something is wrong with the format
-					return checklistChanged, fmt.Errorf("checklist parse error, line %d", l)
+					return checklistChanged, fmt.Errorf("checklist parse error (invalid state), line %d", l)
 				}
 				// check and save comment
-				if checklist.Questions[nextQ-1].Comment != strings.TrimSuffix(commentText, "\n") {
-					checklist.Questions[nextQ-1].Comment = strings.TrimSuffix(commentText, "\n")
+				if checklist.Sections[nextS-1].Questions[nextQ-1].Comment != strings.TrimSuffix(commentText, "\n") {
+					checklist.Sections[nextS-1].Questions[nextQ-1].Comment = strings.TrimSuffix(commentText, "\n")
```
Comment (aleb) [inline input/input.go:283]: We should trim the suffix only once :)

File: input/input.go [hunk 6]
```diff
@@ -241,51 +246,67 @@ func ChecklistEditorInput(repo repository.RepoCommon, checklist bug.Checklist) (
...
+					checklist.Sections[nextS-1].Questions[nextQ-1].State = newState
 					checklistChanged = true
 				}
 				nextQ++
+				if nextQ > len(checklist.Sections[nextS-1].Questions) {
+					nextS++
+					nextQ = 1
+				}
+				if nextS > len(checklist.Sections) {
+					break
```
Comment (aleb) [inline input/input.go:297]: Why is this needed?

File: bug/snapshot.go (hunk 1 above)
Comment (dawzee) [reply bug/snapshot.go:155]: It's because they are two different use cases:
- In GetUserChecklists we're constructing a map of checklists for this user, we want to know if they (and only they) have reviewed it before, if not then we insert the template checklist into the map.
- In GetChecklistCompoundStates we want to know if anyone has reviewed it, if not then the state remains at "Pending"

File: input/input.go (hunk 6 above)
Comment (dawzee) [reply input/input.go:297]: good point. I thought I needed to break out of the loop after the last section/question in case the user added junk after the last status. Just tested without this and it made zero difference, so obviously not =)

File: bug/snapshot.go (hunk 1 above)
Comment (aleb) [reply bug/snapshot.go:155]: I mean we should also check here whether snap.Checklists[lblStr] is present, otherwise we work with a nil map.

Comment (aleb) [reply bug/snapshot.go:155]: Not sure if that is possible, since the snap does have the corresponding label.

Comment (dawzee) [reply bug/snapshot.go:155]: It turns out go let's you check for presence in a nil map without problems. See: https://play.golang.org/p/ARbzTJ8v5E7
